Move GenerateCorrelationId and ValidateCorrelationId to RemoteAuthenticationHandler

PinpointTownes opened the pull request:
Fixes https://github.com/aspnet/Security/issues/647.

/cc @Eilon @Tratcher please don't merge it yet, there are a few things I'm not totally happy with.

dnfclas commented:
Hi __@PinpointTownes__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

PinpointTownes commented on src/Microsoft.AspNetCore.Authentication/RemoteAuthenticationHandler.cs:
```diff
@@ -99,5 +107,69 @@ protected override Task<bool> HandleForbiddenAsync(ChallengeContext context)
+
+            if (!string.Equals(correlationCookie, CorrelationMarker, StringComparison.Ordinal))
+            {
+                Logger.LogWarning(28, "{0} correlation cookie and state property mismatch.", cookieName);
```
This log warning is misleading as the actual check is made earlier, at line 153 (it checks whether a cookie whose name corresponds to the nonce value exists).

This check only ensures the cookie contains "N" :see_no_evil:

PinpointTownes replied:
Options:
* Change the way we generate the cookie to store the generated bits in the cookie value instead of the name (or a few bits in the name and the rest in the value).
* Remove the check, as it's basically useless.
* Move the log message to the appropriate place (i.e L. 158)

@Tratcher @Eilon any preference?

Tratcher replied:
- Don't change the name/format.
- The check is valid, though I don't know what would ever cause the value to be wrong. It's closer to an assert. 
- The log on 158 is adequate for that check.
- Just reword this log to say "The correlation cookie value '{0}' did not match the expected value '{1}'".

PinpointTownes commented on src/Microsoft.AspNetCore.Authentication.OAuth/OAuthHandler.cs:
```diff
@@ -169,7 +166,12 @@ public OAuthHandler(HttpClient backchannel)
+            var properties = new AuthenticationProperties(context.Properties)
+            {
+                // Review: should we add an option to configure the state lifetime?
+                ExpiresUtc = DateTimeOffset.UtcNow + TimeSpan.FromHours(1)
```
Note: a 1-hour lifetime is probably a bit excessive, but it's the default value used by IdentityModel for the OIDC middleware, so I've reused it consistency.

Eilon replied:
BTW you can call `DateTimeOffset.UtcNow.AddHours(1)`.

PinpointTownes replied:
Will do. I'll also use the `SystemClock` property added by @Tratcher in his recent PR to both the OAuth2 and the OIDC middleware.

PinpointTownes replied:
One question remains: should we make it configurable?

PinpointTownes replied:
Updated to use a 15mn lifetime (I just rememberd the OIDC options explicitly replace the default lifetime :smile:)

PinpointTownes commented on global.json:
```diff
@@ -1,3 +1,4 @@
 {
+    "packages": "packages",
```
(I'll remove that when we're done with the other changes)

PinpointTownes commented:
Updated. I fixed the unit tests (it's much easier when you have the packages :smile:)

The `ChallengeSettingState` test still fails. I guess it's caused by a subtle difference in the way `HttpClient` and `UrlEncoder` (used by the fake properties serializer) encode the comma:

`.expires%20Tue,%252026%2520Jan%25202016%252016%253A35%253A26%2520GMT`

`.expires%20Tue%2C%252026%2520Jan%25202016%252016%253A35%253A26%2520GMT`

This test is really cryptic for me. I'm not sure how I'm supposed to fix it...

Tratcher commented:
The test formatter was making assumptions about the data. I was able to get it working by changing AuthenticationPropertiesFormatterKeyValue from
```
-                sb.Append(encoder.Encode(item.Key) + " " + encoder.Encode(item.Value) + " ");
to
+                sb.Append(Uri.EscapeDataString(item.Key) + " " + Uri.EscapeDataString(item.Value) + " ");
```
See https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/KDev/src/Microsoft.IdentityModel.Protocols/AuthenticationProtocolMessage.cs#L113
Give that a try.

PinpointTownes commented:
@Tratcher thanks for the tip! :smile: 

Sadly, I can't test it locally as the xUnit runner no longer works on my machine, even with the latest packages/DNX runtime. Let's see what the CI says :smile:

PinpointTownes commented on src/Microsoft.AspNetCore.Authentication.OAuth/OAuthHandler.cs:
```diff
@@ -177,7 +172,12 @@ public OAuthHandler(HttpClient backchannel)
+            var properties = new AuthenticationProperties(context.Properties)
+            {
+                // Review: should we add an option to configure the state lifetime?
+                ExpiresUtc = Options.SystemClock.UtcNow.AddMinutes(15)
```
So, should we make it configurable?

Tratcher replied:
Yes. It didn't used to expire at all before. `TimeSpan? RoundTripTimeout { get; set; } = TimeSpan.FromMinutes(15);`?

PinpointTownes replied:
The property name is rather clear (though maybe not specific enough?), but I'm not sure we should make it nullable (which would result in a session cookie). At least, not without a warning indicating that setting it to null may result in a terrible cookie flooding :smile:

Tratcher replied:
There wasn't at timeout before in OAuth, but I guess there isn't really a scenario for long timeouts here.

HaoK commented on src/Microsoft.AspNetCore.Authentication/RemoteAuthenticationHandler.cs:
```diff
@@ -2,15 +2,24 @@
     public abstract class RemoteAuthenticationHandler<TOptions> : AuthenticationHandler<TOptions> where TOptions : RemoteAuthenticationOptions
     {
+        private const string CorrelationPrefix = ".AspNetCore.Correlation.";
+        private const string CorrelationProperty = ".xsrf";
```
could make this internal so you don't have so many hardcoded .xsrf's in the tests

PinpointTownes replied:
Yeah, but then we'd have to move these constants to a different class, as this one is generic (it's rather weird/ugly/cumbersome with generics: `RemoteAuthenticationHandler<RemoteAuthenticationOptions>.CorrelationProperty` :trollface:)

HaoK replied:
Don't we have a Constants internal class already?
>

PinpointTownes replied:
AFAICT, nope, no constants class for `Microsoft.AspNetCore.Authentication` (the OAuth2 project has one but this PR removes it as it's now useless).

If you think it's worth it, I'm fine creating one and marking `Microsoft.AspNetCore.Authentication` as a friend assembly to make it accessible from the unit tests project :smile:

HaoK replied:
Nah if we don't have a constants class its not worth adding it for this, hardcoding is fine :)

Tratcher commented on src/Microsoft.AspNetCore.Authentication.OpenIdConnect/OpenIdConnectHandler.cs:
```diff
@@ -184,7 +184,10 @@ public OpenIdConnectHandler(HttpClient backchannel, HtmlEncoder htmlEncoder)
-            var properties = new AuthenticationProperties(context.Properties);
+            var properties = new AuthenticationProperties(context.Properties)
+            {
+                ExpiresUtc = Options.SystemClock.UtcNow.Add(Options.CorrelationLifetime)
```
There's no reason for the correlation lifetime to be independent of the nonce lifetime is there?

Tratcher replied:
Ah, you put CorrelationLifetime on RemoteAuthenticationOptions rather than just OAuthOptions. Hmm..

PinpointTownes replied:
Yep.

In practice, I can't think of a scenario where you'd like to have different lifetimes for the nonce cookie and the correlation cookie (aka `state`).

That said, `state` and `nonce` are really two different things and `nonce` is an OIDC-specific thing, while the antiforgery/correlation cookie is more general and can be applied to any protocol (even to OpenID2, that doesn't support the `state` parameter). In OIDC, using a `nonce` is not mandatory with the code flow and - unlike the correlation check - can be disabled.

So maybe it's a good thing to have a different option for the antiforgery cookie.

Tratcher commented on src/Microsoft.AspNetCore.Authentication/RemoteAuthenticationOptions.cs:
```diff
@@ -56,6 +56,11 @@ public string DisplayName
         public bool SaveTokensAsClaims { get; set; }
 
+        /// <summary>
+        /// Gets or sets the lifetime of the correlation cookie used to prevent cross-site scripting attacks.
```
Yes, but what this really means is that this is the time limit for completing the authentication flow.

Tratcher replied:
That's why I suggested the ridiculous name that I did. Any ideas for a property name that conveys this? There's no reason for this timeout usage to be restricted to the correlation cookie. Some derivations of RemoteAuthenticationHandler might use mechanics other than correlation cookies. E.g. how about applying this to Twitter's State cookie?

HaoK replied:
What was the name you suggested?  Maybe something generic like RemoteAuthenticationTimeout then?

PinpointTownes replied:
If you think `RAT` is clearer, I'm fine :smile: 

I updated the Twitter handler to use the new property. TBH, I'd really like to update it to also use `GenerateCorrelationId`/`ValidateCorrelationId` like the other providers (instead of the custom `RequestToken` and the associated serializer), but I guess it's a bit late for that.

Tratcher replied:
`RemoteAuthenticationTimeout` is good enough.

Yeah, now's not the time to try re-writing Twitter. I think most of that junk was required as part of the OAuth1 or Twitter spec anyways.  Adding the timeout should be sufficient.

PinpointTownes replied:
Updated to use the name you suggested :clap:

Tratcher commented:
:shipit: Squash and rebase?

PinpointTownes commented:
@Tratcher sure. I'll run a few tests to ensure nothing breaks and squash it :smile: 

Any idea why the CI builds fail so hard? I guess it's caused by the DNX -> CLI migration?

HaoK commented:
:fire:  :fire: :fire: :fire: :fire: That pretty much sums up what running tests locally looks like right now as well :fire: :fire:

Tratcher commented:
Yeah, the CLI conversion is causing lots of builds to fail. I can still test locally running build.cmd. Just squash and re-base and I'll run a final test pass.

PinpointTownes commented:
Rebased and squashed :smile: 

I tried the social sample with the Google and GitHub and the OIDC sample with OpenIddict and both worked as intended. I was not able to run the tests locally, tho.

Tratcher commented:
Tested, rebased, and merged.
